src/Lecture2.hs

```diff
@@ -152,12 +183,109 @@ You're free to define any helper functions.
 
 -- some help in the beginning ;)
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
+    { knightHealth    :: Health
+    , knightAttack    :: Attack
+    , knightEndurance :: Endurance
+    } deriving (Eq)
 
-dragonFight = error "TODO"
+data Dragon = Dragon
+    { dragonHealth    :: Health
+    , dragonFirePower :: FirePower
+    } deriving (Eq)
+
+data Chest a = Chest
+    { gold      :: Gold
+    , threasure :: Maybe a
+    } deriving (Show, Eq)
+
+data Reward = Reward Gold (Maybe Treasure) XP deriving (Show, Eq)
+
+data Treasure = Gems | Jewels deriving (Show, Eq)
+
+data EvilDragon = Green Dragon Gold
+                | Black Dragon (Chest Treasure)
+                | Red   Dragon (Chest Treasure) deriving (Eq)
+
+data FightResult = KnightWin Reward | KnightDie | KnightRunAway deriving (Show, Eq)
+
+newtype Gold = Gold Int deriving (Show, Eq)
+
+newtype Health = Health Int deriving (Show, Eq)
+
+newtype Attack = Attack Int deriving (Show, Eq)
+
+newtype Endurance = Endurance Int deriving (Show, Eq)
+
+newtype FirePower = FirePower Int deriving (Show, Eq)
+
+newtype XP = XP Int deriving (Show, Eq)
+
+dragonFight :: Knight -> EvilDragon -> FightResult
+dragonFight = go 1
+    where go :: Int -> Knight -> EvilDragon -> FightResult
+          go strikeNumber knight dragon
+            | strikeNumber > maxStrikesInRound   = go 1 (damageKnight dragon knight) dragon
+            | getDragonHealth dragon <= 0        = KnightWin $ createReward dragon
+            | getKnightEndurance knight <= 0     = KnightRunAway
+            | getKnightHealth knight <= 0        = KnightDie
+            | otherwise                          = go (strikeNumber + 1) (reduceEndurance knight) (damageDragon knight dragon)
+
+maxStrikesInRound :: Int
+maxStrikesInRound = 10
+
+getDragonHealth :: EvilDragon -> Int
+getDragonHealth (Green (Dragon (Health h) _) _) = h
+getDragonHealth (Red (Dragon (Health h) _) _)   = h
+getDragonHealth (Black (Dragon (Health h) _) _) = h
```

Comment:
If you define `Health` as a record newtype, as I suggested in another hint, you'll be able to implement this function using existing functions and function composition without any boilerplate:
```suggestion
getDragonHealth = unHealth . dragonHealth . extractDragon
```

The same is true for other functions where you do nested unwrapping 🎁

Reply:
wow! thank you , really nice!!🥇